### PR TITLE
Add `ToExecuteCommandMatching()` expectation.

### PR DESCRIPTION
**Pull request by jmalloc**
#### What change does this introduce?

This PR adds the `ToExecuteCommandMatching()` expectation, which uses a user-supplied predicate function match command messages.

#### Why make this change?

The existing matches only allow for matching message exactly. Using a predicate function allows for arbitrary matching logic, including the use of third-party matching libraries.

The trade-off is that the failure report is not as accurate. 

#### Is there anything you are unsure about?

No

#### What issues does this relate to?

Partially addresses https://github.com/dogmatiq/testkit/issues/126, `ToRecordCommandMatching()` will be added in a separate PR.

### CODE REVIEW

File: expectation.messagematch.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+// executed that satisfies the given predicate function.
+//
+// Always prefer using ToExecuteCommand() instead, if possible, as it provides
+// more meaningful information in the result of a failure.
+//
+// desc is a human-readable description of the expectation. It should be phrased
+// as an imperative statement, such as "debit the customer".
+//
+// pred is the predicate function. It is called for each executed command. It
+// must return nil at least one for the expectation to pass.
```
Comment (danilvpetrov) [inline expectation.messagematch.go:23]:
```suggestion
// must return nil at least once for the expectation to pass.
```

Comment (koden-km) [inline expectation.messagematch.go:23]: Something not quite right with this last sentence. Should "one" be "once"?
```suggestion
// must return nil at least once for the expectation to pass.
```

**Review by danilvpetrov — APPROVED** [review]
Looking good, just noticed one typo, everything else is quite clear.

File: expectation.messagematch.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+// IgnoreMessage is an error that can be returned by predicate functions used
+// with ToExecuteCommandMatching() and ToRecordEventMatching() to indicate that
+// the predicate does not care about the message and therefore the predicate's
+// result should not affect the expectation's result.
+var IgnoreMessage = errors.New("this message does not need to be inspected by the predicate")
+
+// messageMatchExpectation is an Expectation that checks that a message that
+// satisfies a predicate function is produced.
```
Comment (koden-km) [inline expectation.messagematch.go:44]: Just check that this is clear. Each time I read it I feel like I need to read it again?

Comment (jmalloc) [reply expectation.messagematch.go:44]: extra `that`.

**Review by koden-km — APPROVED** [review]

**Pull request merged**